src/Processors/Formats/Impl/Parquet/ParquetRecordReader.cpp

```diff
@@ -0,0 +1,237 @@
+
+#include <arrow/status.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/column_reader.h>
+#include <parquet/properties.h>
+
+#include "ParquetLeafColReader.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int PARQUET_EXCEPTION;
+}
+
+#define THROW_PARQUET_EXCEPTION(s)                                            \
+    do                                                                        \
+    {                                                                         \
+        try { (s); }                                                          \
+        catch (const ::parquet::ParquetException & e)                         \
+        {                                                                     \
+            auto msg = PreformattedMessage::create("Excepted when reading parquet: {}", e.what()); \
+            throw Exception(std::move(msg), ErrorCodes::PARQUET_EXCEPTION);   \
+        }                                                                     \
+    } while (false)
+
+namespace
+{
+
+std::unique_ptr<parquet::ParquetFileReader> createFileReader(
+    std::shared_ptr<::arrow::io::RandomAccessFile> arrow_file)
+{
+    std::unique_ptr<parquet::ParquetFileReader> res;
+    THROW_PARQUET_EXCEPTION(res = parquet::ParquetFileReader::Open(std::move(arrow_file)));
+    return res;
+}
+
+std::unique_ptr<ParquetColumnReader> createColReader(
+    const parquet::ColumnDescriptor & col_descriptor,
+    DataTypePtr ch_type,
+    std::unique_ptr<parquet::ColumnChunkMetaData> meta,
+    std::unique_ptr<parquet::PageReader> reader)
+{
+    if (col_descriptor.logical_type()->is_date() && parquet::Type::INT32 == col_descriptor.physical_type())
+    {
+        return std::make_unique<ParquetLeafColReader<ColumnInt32>>(
+            col_descriptor, std::make_shared<DataTypeDate32>(), std::move(meta), std::move(reader));
+    }
+    else if (col_descriptor.logical_type()->is_decimal())
+    {
+        switch (col_descriptor.physical_type())
+        {
+            case parquet::Type::INT32:
+            {
+                auto data_type = std::make_shared<DataTypeDecimal32>(
+                    col_descriptor.type_precision(), col_descriptor.type_scale());
+                return std::make_unique<ParquetLeafColReader<ColumnDecimal<Decimal32>>>(
+                    col_descriptor, data_type, std::move(meta), std::move(reader));
+            }
+            case parquet::Type::INT64:
+            {
+                auto data_type = std::make_shared<DataTypeDecimal64>(
+                    col_descriptor.type_precision(), col_descriptor.type_scale());
+                return std::make_unique<ParquetLeafColReader<ColumnDecimal<Decimal64>>>(
+                    col_descriptor, data_type, std::move(meta), std::move(reader));
+            }
+            case parquet::Type::FIXED_LEN_BYTE_ARRAY:
+            {
+                if (col_descriptor.type_length() <= static_cast<int>(sizeof(Decimal128)))
```

Comment:
Indeed, for parquet file with `Decimal32` type generated with ClickHouse, the reading result is correct for current codes, while it seems a little inefficient. And the reading logic in arrow is same like this, there will not be too much optimization for current logic, because the value of `type_length` can be 3 bytes for `Decimal32`.

Further more, for the parquet file generated by Spark, the `physical_type` of `Decimal32` is `INT32` instead of `FIXED_LEN_BYTE_ARRAY`, both the reading and writing performance will be better. So, I prefer to change the write logic of ClickHouse instead of optimizing for a bad choice.